minimalkv/fsspecstore.py

```diff
@@ -0,0 +1,209 @@
+        self.mkdir_prefix = mkdir_prefix
+
+    @lazy_property
+    def _prefix_exists(self):
+        # Check if prefix exists.
+        # Used by inheriting classes to check if e.g. a bucket exists.
+        try:
+            self._fs.info(self.prefix)
+        except (FileNotFoundError, IOError):
+            return False
+        return True
+
+    def iter_keys(self, prefix: str = "") -> Iterator[str]:
+        """Iterate over all keys in the store starting with prefix.
+
+        Parameters
+        ----------
+        prefix: str, optional, default = ''
+            Only iterate over keys starting with prefix. Iterate over all keys if empty.
+
+        Raises
+        ------
+        IOError
+            If there was an error accessing the store.
+        """
+        # List files
+        all_files_and_dirs = self._fs.find(f"{self.prefix}", prefix=escape(prefix))
+
+        return map(
+            lambda k: unescape(k.replace(f"{self.prefix}", "")), all_files_and_dirs
+        )
+
+    def _delete(self, key: str):
+        try:
+            self._fs.rm_file(f"{self.prefix}{escape(key)}")
+        except FileNotFoundError:
+            pass
+
+    def _open(self, key: str) -> IO:
+        try:
+            return self._fs.open(f"{self.prefix}{escape(key)}")
+        except FileNotFoundError:
+            raise KeyError(key)
+
+    def _put_file(self, key: str, file: IO) -> str:
+        self._fs.pipe_file(f"{self.prefix}{escape(key)}", file.read())
+        return key
+
+    def _has_key(self, key):
+        return self._fs.exists(f"{self.prefix}{escape(key)}")
+
+    def _create_filesystem(self):
+        # To be implemented by inheriting classes.
+        raise NotImplementedError
+
+    @lazy_property
+    def _fs(self) -> AbstractFileSystem:
+        fs = self._create_filesystem()
+
+        if self.mkdir_prefix:
+            fs.mkdir(self.prefix)
+        return fs
+
+    # Skips lazy properties.
+    # These will be recreated after unpickling through the lazy_property decorator
+    def __getstate__(self):  # noqa D
+        return {
+            key: value
+            for key, value in self.__dict__.items()
+            if not key.startswith(LAZY_PROPERTY_ATTR_PREFIX)
+        }
+
+
+def escape(path: str) -> str:
+    """
+    URL-escape a path.
+
+    Every character not allowed within a URL segment is escaped, including slashes.
+    This is necessary because gcsfs does not escape the path properly.
+
+    Parameters
+    ----------
+    path: str
+        The path to escape.
+
+    Returns
+    -------
+    str
+        The escaped path.
+    """
+    return quote(path, safe="")
+
+
+def unescape(path: str) -> str:
+    """
+    URL-unescape a path. Every character not allowed within a URL segment is unescaped, including slashes.
+
+    Parameters
+    ----------
+    path: str
+        The path to unescape.
+
+    Returns
+    -------
+    str
+        The unescaped path.
+    """
+    return unquote(path)
```

Comment:
why wrap these simple function calls?

Reply:
i see, that makes sense. still, maybe a bit too much boilerplate for something so simple.

how do you like:

```py
from functools import partial
from urllib.parse import quote as _quote
from urllib.parse import unquote

quote = partial(_quote, safe="")
```

and then using `quote` and `unquote` instead of `escape` and `unescape`?